examples/minimal_pub_sub/src/publisher_member_function.rs

```diff
@@ -0,0 +1,59 @@
+use std::{
+    env,
+    sync::{Arc, Mutex},
+    thread,
+    time::Duration,
+};
+
+use anyhow::Result;
+
+struct Publisher {
+    publisher: Arc<Mutex<rclrs::Publisher<std_msgs::msg::String>>>,
+    publish_count: Arc<Mutex<u32>>,
+}
+
+unsafe impl Send for Publisher {}
+
+impl Publisher {
+    pub fn new(context: &rclrs::Context) -> Self {
+        let node = rclrs::create_node(context, "publisher").unwrap();
+
+        let publisher = node
+            .create_publisher::<std_msgs::msg::String>("topic", rclrs::QOS_PROFILE_DEFAULT)
+            .unwrap();
+
+        Self {
+            publisher: Arc::new(Mutex::new(publisher)),
+            publish_count: Arc::new(Mutex::new(0)),
+        }
+    }
+
+    fn init(&mut self) {
+        let publish_count = self.publish_count.clone();
+        let publisher = self.publisher.clone();
+
+        thread::spawn(move || loop {
+            thread::sleep(Duration::from_secs(1));
+
+            let msg = std_msgs::msg::String {
+                data: format!("Hello, world! {}", publish_count.lock().unwrap()),
+            };
+
+            println!("Publishing: [{}]", msg.data);
+
+            publisher.lock().unwrap().publish(msg).unwrap();
+            let mut num_count = publish_count.lock().unwrap();
+            *num_count += 1;
+        });
+    }
+}
+
+fn main() -> Result<(), rclrs::RclrsError> {
+    let context = rclrs::Context::new(env::args())?;
+    let mut publisher = Publisher::new(&context);
+    publisher.init();
+    while context.ok() {
+        std::thread::sleep(std::time::Duration::from_millis(100));
+    }
+    Ok(())
+}
```

Comment:
Here's what I would change:
* add a timer_callback() like in the original rclcpp version
* use an atomic variable instead of a Mutex for the counter
* make all fallible methods return Result
* remove unneeded lines (e.g. `unsafe impl Send for Publisher {}` and `use anyhow::Result`)


```suggestion
use std::{
    env,
    sync::atomic::{AtomicU32, Ordering},
    sync::{Arc, Mutex},
    thread,
    time::Duration,
};

use rclrs::RclrsError;

struct MinimalPublisher {
    publisher: Arc<Mutex<rclrs::Publisher<std_msgs::msg::String>>>,
    publish_count: AtomicU32,
}

impl MinimalPublisher {
    pub fn new(context: &rclrs::Context) -> Result<Self, RclrsError> {
        let node = rclrs::create_node(context, "publisher")?;

        let publisher =
            node.create_publisher::<std_msgs::msg::String>("topic", rclrs::QOS_PROFILE_DEFAULT)?;

        Ok(Self {
            publisher: Arc::new(Mutex::new(publisher)),
            publish_count: AtomicU32::new(0),
        })
    }

    fn timer_callback(&self) -> Result<(), RclrsError> {
        let cur_publish_count = self.publish_count.fetch_add(1, Ordering::Relaxed);
        let msg = std_msgs::msg::String {
            data: format!("Hello, world! {}", cur_publish_count),
        };

        println!("Publishing: [{}]", msg.data);
        self.publisher.lock().unwrap().publish(msg)
    }

    // Simulates a timer
    fn run(&self) -> Result<(), RclrsError> {
        loop {
            thread::sleep(Duration::from_secs(1));
            self.timer_callback()?;
        }
    }
}

fn main() -> Result<(), rclrs::RclrsError> {
    let context = rclrs::Context::new(env::args())?;
    let publisher = MinimalPublisher::new(&context)?;
    publisher.run()
}
```

Reply:
That said, I'm on the fence on whether adding this example makes sense before we have timers. With timers, it may look very different.